packages/patrol_cli/lib/src/coverage/run_code_coverage.dart

```diff
@@ -0,0 +1,190 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:coverage/coverage.dart';
+import 'package:glob/glob.dart';
+import 'package:patrol_cli/src/base/logger.dart';
+import 'package:patrol_cli/src/devices.dart';
+import 'package:vm_service/vm_service.dart';
+import 'package:vm_service/vm_service_io.dart';
+
+Future<Map<String, HitMap>> _collectCoverage(
+  VmService client,
+  Uri vmUri,
+  String packageName,
+  String mainIsolateId,
+) async {
+  final coverage = await collect(
+    vmUri,
+    false,
+    false,
+    false,
+    {packageName},
+  );
+
+  final socket = await WebSocket.connect(client.wsUri!)
+    ..add(
+      jsonEncode(
+        {
+          'jsonrpc': '2.0',
+          'id': 21,
+          'method': 'ext.patrol.markTestCompleted',
+          'params': {
+            'isolateId': mainIsolateId,
+            'command': 'markTestCompleted',
+          },
+        },
+      ),
+    );
+  await socket.close();
+
+  final map = await HitMap.parseJson(
+    coverage['coverage'] as List<Map<String, dynamic>>,
+  );
+
+  return map;
+}
+
+Future<ProcessResult> _forwardAdbPort(String host, String guest) async {
+  return Process.run('adb', ['forward', 'tcp:$host', 'tcp:$guest']);
+}
+
+Uri _createWebSocketUri(Uri uri) {
+  final pathSegments = uri.pathSegments.where((c) => c.isNotEmpty).toList()
+    ..add('ws');
+  return uri.replace(scheme: 'ws', pathSegments: pathSegments);
+}
+
+Future<void> _saveCoverage(String report) async {
+  final coverageDirectory = Directory('coverage');
+
+  if (!coverageDirectory.existsSync()) {
+    await coverageDirectory.create();
+  }
+  await File(
+    coverageDirectory.uri.resolve('patrol_lcov.info').toString(),
+  ).writeAsString(report);
+}
+
+Future<void> runCodeCoverage({
+  required String flutterPackageName,
+  required Directory flutterPackageDirectory,
+  required TargetPlatform platform,
+  required Logger logger,
+  required Set<Glob> ignoreGlobs,
+}) async {
+  final homeDirectory =
+      Platform.environment['HOME'] ?? Platform.environment['USERPROFILE'];
+
+  final logsProcess = await Process.start(
```

Comment:
I think a `DisposeScope` should be passed here, and when it is disposed, processes should be killed and subscriptions canceled. This is a general pattern in this repo - see [AndroidTestBackend](https://github.com/leancodepl/patrol/blob/9f97af3615374ba8d5996e5ba6ad75496d72b7c5/packages/patrol_cli/lib/src/android/android_test_backend.dart#L29) as an example.